### PR TITLE
[Security] Rename User to InMemoryUser

### LINKED ISSUE #26348: [RFC][Security] Simplify the User and UserChecker
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | 4.1

In #23508 the `AdvancedUserInterface` was deprecated and the `UserChecker` was made to support the `User` specifically. When 5.0 is released, that means that the `UserChecker` will effectively _only_ work when the `User` is passed.

https://github.com/symfony/symfony/blob/97cee3a157cf480a9204fa45816e5d76db32df72/src/Symfony/Component/Security/Core/User/UserChecker.php#L31-L37

Some facts:
 - The `User` class is not used anywhere in the core except for the `LdapUserProvider` and `InMemoryUserProvider`.
 - The `InMemoryUserProvider` only supports the `enabled` feature of the `AdvancedUserInterface`.
 - The `LdapUserProvider` doesn't use any of the `AdvancedUserInterface` features.

## RFC

In this RFC I would like to propose the following things to reduce the complexity:
 - Create an `LdapUser`, exclusively used by the `LdapUserProvider`. For BC it can extend the User for instanceof checks. Would only receive the currently used fields.
 - Create an `InMemoryUser`, exclusively used by the `InMemoryUserProvider`. For BC it can extend the User for instanceof checks. Would only receive the currently used fields.
 - Add a `NullUserChecker`, which features 0 checks. In 5.0 this would become the default `UserChecker` in the configuration.
 - Rename the current `UserChecker` to `InMemoryUserChecker` (BC: `UserChecker extends InMemoryUserChecker`). This UserChecker will no longer check anything but the `enabled` flag. Keep this user-checker as default in 4.x, but throw a deprecation if no user-checker is configured.
 - Deprecated the current `User` user class and instead document thoroughly how to create your own user class _and_ checker.
 - Provide an easy-to-use `UserInterface` implementation for tests only.

## Scenarios
I've been thinking about a few scenarios in which things will change for developers. Generally speaking, the impact should be rather small.

### Using the `User` in custom `UserProvider`
_Time to fix: roughly 15~30 minutes, requires proper documentation_

In this scenario, the develop will have to carefully look at which features should be kept. A custom `UserInterface` implementation will have to be provided. If any of the `AdvancedUserInterface` domain-like features were used, these will have to be manually implemented, and a custom `UserChecker` has to be created and configured.

To do:
 - Write custom user class and configure it in the provider that would previously create a `User`
 - Write custom user checker and configure it in the firewall

### Using the `InMemoryUser` and `InMemoryUserProvider`
_Time to fix: roughly 2 minutes, proper deprecation should suffice_

In this scenario, by default, the `UserChecker` would be configured by Symfony as default config. In 5.0 this would become the `NullUserChecker`, which does nothing. The `InMemoryUser` would be able to define an `enabled` field, which has to be checked by the `InMemoryUserChecker`. To fix this, configure the right user checker

To do:
 - Configure the `user-checker` option in your firewall to use the `InMemoryUserChecker`

### Using the `User` in tests
_Time to fix: roughly 5~30 minutes, depending on how often it's used_

In this scenario, which I believe is more common, developers have used the `User` in their tests. This is also the case within Symfony. I think that it would be easiest to add a `TestUser` in a `Test` namespace and expose this (as opposed to tests in the `Tests` namespace).

To do:
 - Replace the usages of `User` by `TestUser` with the right namespace 
---

If I've missed any scenarios, please let me know and I will add them. The goal of this RFC is to reduce domain logic in Symfony and having no ambiguous purposes for the User and User Checker. Additionally the `NullUserChecker` will simply do nothing, which is the default case with a custom user class as of 4.1.

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | Closes #26348
| License       | MIT
| Doc PR        | -

This PR aims to clarify that the `User` class should only be used by the `InMemoryUserProvider`, as documented:
https://github.com/symfony/symfony/blob/c06a76c38410ac027abce10a68bc219add7bf07d/src/Symfony/Component/Security/Core/User/User.php#L15-L17
It also renames `UserChecker` to `InMemoryUserChecker` because it only works with the in-memory user class:
https://github.com/symfony/symfony/blob/c06a76c38410ac027abce10a68bc219add7bf07d/src/Symfony/Component/Security/Core/User/UserChecker.php#L31-L32

### CODE REVIEW

**Comment by chalasr** [comment]
The travis failure with deps=high is expected. This is ready

File: UPGRADE-6.0.md [hunk 1]
```diff
@@ -174,6 +174,8 @@ Routing
 Security
 --------
 
+ * Deprecate class `User`, use `InMemoryUser` or your own implementation instead
```
Comment (rosier) [inline UPGRADE-6.0.md]:
```suggestion
 * Remove class `User`, use `InMemoryUser` or your own implementation instead
```

File: UPGRADE-6.0.md [hunk 2]
```diff
@@ -174,6 +174,8 @@ Routing
...
+ * Deprecate class `UserChecker`, use `InMemoryUserChecker` or your own implementation instead
```
Comment (rosier) [inline UPGRADE-6.0.md]:
```suggestion
 * Remove class `UserChecker`, use `InMemoryUserChecker` or your own implementation instead
```

File: src/Symfony/Component/Security/Core/User/InMemoryUser.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+{
+    private $username;
+    private $password;
+    private $enabled;
+    private $roles;
+
+    /**
+     * @param string[] $roles
+     */
+    public function __construct(?string $username, ?string $password, array $roles = [], bool $enabled = true)
```
Comment (rosier) [inline src/Symfony/Component/Security/Core/User/InMemoryUser.php]:
```suggestion
    public function __construct(string $username, ?string $password, array $roles = [], bool $enabled = true)
```
Or is that not possible for BC ?

Comment (chalasr) [reply src/Symfony/Component/Security/Core/User/InMemoryUser.php]: It is possible :) I need to double check that there is no legit use case for `null`, but 👍 . Thanks @rosier!

File: src/Symfony/Component/Security/Core/User/User.php [hunk 4]
```diff
@@ -11,12 +11,16 @@
 
 namespace Symfony\Component\Security\Core\User;
 
+trigger_deprecation('symfony/security-core', '5.3', 'The "%s" class is deprecated, use "%s" instead.', User::class, InMemoryUser::class);
```
Comment (noniagriconomie) [inline src/Symfony/Component/Security/Core/User/User.php:14]: what about also telling that the class will be removed in v6.0?
according to the comment https://github.com/symfony/symfony/pull/40443/files#r592508122

Comment (fabpot) [reply src/Symfony/Component/Security/Core/User/User.php:14]: We don't do that anymore. Anything deprecated in 5.x is going to be removed in 6.0, so being explicit is not needed.

Comment (noniagriconomie) [reply src/Symfony/Component/Security/Core/User/User.php:14]: Understood, I was seeing sometime that is why :)

**Comment by chalasr** [comment]
Thanks for the review @rosier, all comments have been addressed.

File: src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+use Symfony\Component\Security\Core\Exception\AccountExpiredException;
+use Symfony\Component\Security\Core\Exception\CredentialsExpiredException;
+use Symfony\Component\Security\Core\Exception\DisabledException;
+use Symfony\Component\Security\Core\Exception\LockedException;
+
+/**
+ * Checks the state of the in-memory user account.
+ *
+ * @author Robin Chalas <robin.chalas@gmail.com>
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php]: This class is a copy/paste of UserChecker.php, right?

File: src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            return;
+        }
+
+        if (!$user->isEnabled()) {
+            $ex = new DisabledException('User account is disabled.');
+            $ex->setUser($user);
+            throw $ex;
+        }
+
+        // @deprecated since Symfony 5.3
```
Comment (fabpot) [inline src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php:40]: What about NOT supporting the deprecated stuff in the new class?

Comment (chalasr) [reply src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php:40]: Given this makes the `security.user_checker` service use the new class, not supporting those checks would be a behavior change. As this is about security and all is done via configuration (you don't need to manipulate those classes to use them), I think it's worth keeping those checks until 6.0

File: src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php (hunk 5 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/User/InMemoryUserChecker.php]: right :)

File: src/Symfony/Component/Routing/Tests/Fixtures/AttributesFixtures/AttributesClassParamAfterCommaController.php [hunk 7]
```diff
@@ -3,14 +3,13 @@
...
 use Symfony\Component\Routing\Tests\Fixtures\Attributes\FooAttributes;
-use Symfony\Component\Security\Core\User\User;
 
 #[FooAttributes(
     foo: [
         'bar' => ['foo','bar'],
         'foo'
     ],
-    class: User::class
+    class: \stdClass::class
```
Comment (derrabus) [inline src/Symfony/Component/Routing/Tests/Fixtures/AttributesFixtures/AttributesClassParamAfterCommaController.php]: We might want to backport those fixture changes.

File: src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php [hunk 8]
```diff
@@ -85,6 +86,9 @@ public function testSerializeNullImpersonateUrl()
         $this->assertNull($unserializedToken->getOriginatedFromUri());
     }
 
+    /**
+     * @group legacy
```
Comment (derrabus) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Token/SwitchUserTokenTest.php]: The fixture should be updated instead.

File: src/Symfony/Component/Security/Core/User/UserChecker.php [hunk 9]
```diff
@@ -16,13 +16,24 @@
...
+ *
+ * @deprecated since Symfony 5.3, use {@link InMemoryUserChecker} instead
  */
 class UserChecker implements UserCheckerInterface
 {
+    private $memoryChecker;
+
+    public function __construct()
+    {
+        $this->memoryChecker = new InMemoryUserChecker();
```
Comment (derrabus) [inline src/Symfony/Component/Security/Core/User/UserChecker.php]: You assign this variable, but never use it.

File: src/Symfony/Component/Routing/Tests/Fixtures/AttributesFixtures/AttributesClassParamAfterCommaController.php (hunk 7 above)
Comment (chalasr) [reply src/Symfony/Component/Routing/Tests/Fixtures/AttributesFixtures/AttributesClassParamAfterCommaController.php]: done in #40463

File: src/Symfony/Component/Security/Core/Tests/User/InMemoryUserTest.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        return [
+            [true, new InMemoryUser('username', 'password'), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('username', 'password', ['ROLE']), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('username', 'password', ['ROLE']), new InMemoryUser('username', 'password', ['NO ROLE'])],
+            [false, new InMemoryUser('diff', 'diff'), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('diff', 'diff', [], false), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('diff', 'diff', [], false, false), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('diff', 'diff', [], false, false, false), new InMemoryUser('username', 'password')],
+            [false, new InMemoryUser('diff', 'diff', [], false, false, false, false), new InMemoryUser('username', 'password')],
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Tests/User/InMemoryUserTest.php]: These last 3 tests can be removed, `InMemoryUser` removed the last 3 constructor arguments.

File: src/Symfony/Component/Security/Core/Tests/User/InMemoryUserCheckerTest.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $checker = new InMemoryUserChecker();
+        $this->assertNull($checker->checkPostAuth(new InMemoryUser('John', 'password')));
+    }
+
+    public function testCheckPreAuthDisabled()
+    {
+        $this->expectException(DisabledException::class);
+        $checker = new InMemoryUserChecker();
+        $checker->checkPreAuth(new InMemoryUser('John', 'password', [], false, true, false, true));
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Tests/User/InMemoryUserCheckerTest.php]:
```suggestion
        $checker->checkPreAuth(new InMemoryUser('John', 'password', [], false));
```

`InMemoryUser` doesn't have these extra arguments

File: src/Symfony/Component/Security/Core/User/UserChecker.php [hunk 12]
```diff
@@ -16,10 +16,14 @@
...
 use Symfony\Component\Security\Core\Exception\LockedException;
 
+trigger_deprecation('symfony/security-core', '5.3', 'The "%s" class is deprecated, use "%s" instead.', UserChecker::class, InMemoryUserChecker::class);
+
 /**
  * UserChecker checks the user account flags.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @deprecated since Symfony 5.3, use {@link InMemoryUserChecker} instead
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/User/UserChecker.php]: Shouldn't we instead use `class_alias`, like here:
https://github.com/symfony/symfony/blob/5.x/src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpFactory.php#L16
https://github.com/symfony/symfony/blob/5.x/src/Symfony/Component/Messenger/Bridge/Amqp/Transport/AmqpFactory.php#L36

File: src/Symfony/Component/Security/CHANGELOG.md [hunk 13]
```diff
@@ -4,6 +4,8 @@ CHANGELOG
 5.3
 ---
 
+ * Deprecate class `User`, use `InMemoryUser` instead
```
Comment (wouterj) [inline src/Symfony/Component/Security/CHANGELOG.md:7]: I think we should be more explicit that this is not a drop-in replacement, as methods related to advanced user functionality have been removed

**Review by wouterj — APPROVED** [review]
Thanks Robin!

Am I missing something, or are the 2 `src/Symfony/Component/Security/Core/Tests/Authentication/Token/Fixtures/switch-user-token-*.txt` files not used?

**Review by wouterj — CHANGES_REQUESTED** [review]
Sorry, I selected the wrong choice :facepalm:

File: src/Symfony/Component/Security/CHANGELOG.md (hunk 13 above)
Comment (chalasr) [reply src/Symfony/Component/Security/CHANGELOG.md:7]: Done in UPGRADE files

**Comment by chalasr** [comment]
> Am I missing something, or are the 2 src/Symfony/Component/Security/Core/Tests/Authentication/Token/Fixtures/switch-user-token-*.txt files not used?

Only one of two, actually! It's used in SwitchUserListenerTest.

Thanks for the review Wouter, all fixed.

**Review by wouterj — APPROVED** [review]

**Comment by fabpot** [comment]
Are the broken tests related to the changes here?

**Comment by chalasr** [comment]
@fabpot yes but only deps=high fails, I think it will be fixed once merged. 
In case I'm wrong, be sure that I'll send a PR to fix it right away.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**